Add prototype cheap optimizer

fritzo opened the pull request:
Addresses #56 
Mirrors ~~https://github.com/uber/pyro/pull/1424~~ https://github.com/uber/pyro/pull/1430

## Description

Implements a cheap approximate optimizer for very large paths. This is currently over 100x faster on some large contractions in Pyro, but produces slightly suboptimal paths.

## Todos
- [x] Fix issues with determinism, required by `shared_intermediates`
- [x] Ensure algorithm is not much worse than `greedy`
- ~~Add quality tests (will be easier after #61)~~

## Questions
- [x] How should we be computing the final outer products? Can we do this all in one go?
- [x] What tests should I add?

## Status
- [x] Ready to go

fritzo commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+        k2s = set(k2 for dim in k1 for k2 in dim_to_keys[dim] if k2 != k1)
+        _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
+
+    # Greedily compute pairwise outer products.
```
@jcmgray Am I doing this right? I don't really understand the outer product phase, and I'm not sure what the difference is between this and the simpler
```py
# Perform a final outer-product contraction.
ssa_path.append(tuple(remaining.values()))
return ssa_path
```

jcmgray replied:
The difference is just efficiency, although I agree it's quite a niche scenario. Doing pairwise outer products is generally faster than performing the entire outer product simultaneously with einsum (and whats more permits non-einsum backends). There's a test for this: ``test_can_optimize_outer_products``. 

I think you should be able to use the exact same costing/logic as the inner products, the only difference being that we only invoke checking all the outer products if there >1 remaining tensors but no remaining inner products (this is an approximate heuristic - sometimes the optimal path involves performing outer products first).

fritzo replied:
Great, I've added tests cases to `test_can_optimize_outer_products`, refined the objective function, and removed the extraneous final contraction (moved to top of function as an edge case).

> use the exact same costing/logic as the inner products

Note the one difference: this phase optimizes for small inputs `(x1, x2)` independently, rather than optimizing for small outputs `x12`. By using a separable objective function, this phase has linear complexity; whereas if we used the same logic as for contractions this phase would have quadratic complexity. I'm not sure how much this matters in practice.

jcmgray replied:
Nice. Yeah if you are only left with outer products there should only be a few of them since they produce an exponentially large tensor (the most extreme case for example would be about <~30 tensors each of size 2). Your approach looks good however!

fritzo replied:
So I just realized that this is an important case, e.g. if we have many pairs of contractions all lumped into a single einsum, e.g. "a,a,b,b,c,c,d,d,e,e". This results in linearly-many outer products (i.e. multiplies). So it's at least a little important to avoid quadratic growth here.

jcmgray commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+        keys = list(keys)
+        for i, k1 in enumerate(keys):
+            k2s = keys[:i]
+            _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
```
Still wrapping my head around this part a bit. Could you not just iterate over each ``dim`` that appears more than twice (``dim_ref_counts[2]``)?

I think I see that that you only ever need to add the minimum cost from ``k1`` between ``k2s``, is this the reason for not just pushing all the pairwise candidates?

fritzo replied:
Originally I pushed all the candidates, but ended up with something like 30000 candidates in my test case. This ended up slowing down the O(n log(n)) heap logic. Switching to eager O(n) minimization gave me about a 10x speedup on my large test case.

jcmgray commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+    either = k1 | k2
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
```
I like this way of keeping track of which indices should be maintained, just a note that it would be good to use the same logic in `optimal` eventually.

fritzo replied:
Yeah `_get_candidate()` is the hotspot in this algorithm, actually this line was the hottest spot. I got about 2x speedup by switching from for loops to `frozenset` logic here.

jcmgray commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+        _, ssa_id1, k1 = heapq.heappushpop(queue, (cost, ssa_id12, k12))
+
+    # Perform one final reduction to match output shape.
+    ssa_path.append((ssa_id1,))
```
Yes see comment above about output indices - I think this shouldn't be necessary.

jcmgray commented:
I've updated a gist I have of contraction finding time [here](https://gist.github.com/jcmgray/2fb88f1b8473eef5070316f02e26d66e), it looks good for large contractions!

![index](https://user-images.githubusercontent.com/8982598/46289199-417d0c00-c580-11e8-9153-fcfdd6e33324.png).

In a 'real-world' contraction, there is a slight difference in timings, maybe due to ignoring FLOP cost? Might also just be random though.

fritzo commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+    """
+    # A dim common to all terms might as well be an output dim.
+    inputs = list(map(frozenset, inputs))
+    output = frozenset(output) | frozenset.intersection(*inputs)
```
I'll expand on this comment. This is purely for performance. The issue is that the overall algorithm is quadratic in the number of tensors that share any dimension. Often all tensors share a batch dimension. But the path we produce should be the same whether or not the problem is batched, so I assume wlog that we're not reducing along this dimension. This led to a >10x speedup in my large test case.

jcmgray replied:
Ah I think I understand now - so if you don't do this, any broadcast indices make the graph maximally connected. I don't generally have any broadcast indices in my contractions thus this is a bit of an alien situation. If you have a typical one to hand as an example that might be nice for me to reason/experiment with but no worries if not!

fritzo replied:
Yes this is a very common example in Pyro. Here is a large [test case](https://github.com/uber/pyro/blob/9a10d0e/tests/ops/einsum/test_paths.py#L171). Alternatively, simply append `z` to all inputs in any of your standard test cases :smile:

jcmgray replied:
Thanks, and yes of course - but it wouldn't be realistic :wink:.

dgasmith replied:
Did you find that `frozenset` was more performant than set overall? We have a number of places where we could replace `set` with `frozenset` if true.

fritzo replied:
I think `frozenset` is equally performant with `set`. What I found was that `frozenset` was more performant than manually computing intersections via a for loop. The main reason I'm using `frozenset`s here is to be able to use the sets as keys of dictionaries. But I also find immutable data structures make it easier to reason about code (tuples over lists, frozensets over sets).

dgasmith commented on opt_einsum/tests/test_paths.py:
```diff
@@ -172,10 +176,12 @@ def test_greedy_edge_cases():
     d = np.random.randn(10, 2)
-    assert oe.contract_path("ab,cd,ef,fg", a, b, c, d, path='greedy')[0] == [(2, 3), (0, 2), (0, 1)]
+    actual = oe.contract_path("ab,cd,ef,fg", a, b, c, d, path=optimize)[0]
+    assert actual in [[(2, 3), (0, 2), (0, 1)], [(2, 3), (0, 2), (0, 1)]]
```
The righthand side elements look identical here?

jcmgray commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,201 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
+    cost = helpers.compute_size_by_dict(k12, sizes)
```
What was the reasoning for moving away from ``size(k12) - size(k1) - size(k2)``? It produces noticeable worse contractions for me in quite a few examples.

fritzo replied:
Oh ok, I'll change back to a variant of `size12 - size1 - size2`. My motivation was that this objective led to more determinism in a large example I was looking at, and the increased sharing led to overall cheaper cost. I'll see if I can achieve the same effect with a different tie-breaker.

jcmgray replied:
Yes for most 'difficult' examples I've tried (e.g. contracting quantum circuits), doing it this way produces contractions with a bit worse to a lot worse scaling so my preference would be to keep it.

On a side note, it might be nice at some point to be able to supply a custom function here (e.g. just ``size12``). I particularly like the idea of supplying several even intentionally random costs and then finding the best scaling across several attempts. Might be an issue best left for later however!

dgasmith replied:
Echoing @jcmgray here, if this isn't too much of a hotspot `size(k12) - size(k1) - size(k2)` should prevent some obviously poor pathways.

jcmgray commented:
Do we want this to replace the current ``'greedy'`` method? I think it probably could, and as long as it maintains the ``size12 - size1 - size2`` cost it should find essentially the same contraction paths (modulo a bit of variance from choosing different branches when several have the same cost).

The one remaining difference (I think?) is that the current 'greedy' breaks ties using the flop-cost, but adding this back in might make most sense using a custom cost function like proposed in #64.

We could keep the current greedy around as ``'greedy_legacy'`` or something just for reference.

fritzo commented on opt_einsum/contract.py:
```diff
@@ -349,6 +353,8 @@ def contract(*operands, **kwargs):
         - 'optimal' An algorithm that tries all possible ways of
           contracting the listed tensors. Scales exponentially with
           the number of terms in the contraction.
+        - 'cheap' An approximate algorithm that is cheaper than 'greedy'
```
Ha ha no love fore "cheap" :smile: Sure, I'll change to "eager".

fritzo commented:
Ok, I've addressed review comments. The current cost function is `size12 - (size1 + size2) / 2`, since this seems to work on the few sharing examples I've tried. I don't mind if you change it back to `size12 - size1 - size2`.

I think it's reasonable to merge (pending further comments) and then build some experimentation tooling to determine what objective function works will. One thing that would be handy would be a FLOP counter that worked with sharing; currently I'm running an actual computation and counting number of things in the sharing `cache`.

jcmgray commented on opt_einsum/paths.py:
```diff
@@ -272,3 +279,210 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+    two = k1 & k2
+    one = either - two
+    k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
+    cost = costs[k1] + costs[k2] + helpers.compute_size_by_dict(k12, sizes)
```
I think I would prefer two things here:

1. ``size12 - size1 - size2`` I still think is the best default, it works best for all the tricky contractions I have (current version with ``(size1 -size2) / 2`` varies between 1-4x worse), and is quite natural since it is just proportional to the amount of memory removed.

2. Maybe this was done for efficiency, but I think its a bit confusing having the cost logic spread out over several places, i.e. ``costs[k1]`` and ``costs[k1]`` are halved and negative here, but that's not very clear. If the line was explicitly ``cost = footprint[k12] - footprint[1] - footprint[2]`` I think that would be beneficial, particularly if it comes to implementing a custom cost function.

fritzo replied:
Done, thanks for reviewing. I believe this should now be very close to the `greedy` implementation.

fritzo commented on opt_einsum/contract.py:
```diff
@@ -42,23 +42,23 @@ def __repr__(self):
-            "-" * 80 + "\n",
-            "{:>6} {:>11} {:>22} {:>37}\n".format(*header),
-            "-" * 80
+            u"  Complete contraction:  {}\n".format(overall_contraction),
```
This fixes a Python 2 unicode bug in `PathInfo`. I could pull this out into a separate PR if you want.

jcmgray replied:
Fine at least for me to leave this in, thanks for fixing it.

jcmgray commented:
All looks great to me. I've updated this gist of [timings and other tests here](https://gist.github.com/jcmgray/2fb88f1b8473eef5070316f02e26d66e) in case its of interest (main timings similar to above still).

fritzo commented:
@jcmgray @dgasmith Thanks for the extensive review!
